rfcs/20200424-sparse-domain-isolation.md

```diff
@@ -0,0 +1,620 @@
+    wich is similar to tf.nn.embedding_lookup_sparse in function.
+
+  Args:    
+     All args ars similar to `tf.nn.embedding_lookup_sparse`.
+  Returns:
+    A `dynamic_embedding.TrainableWarp` which hold the handles of `params` and `ids`.
+  """
+  pass
+```
+
+### APIs of training sparse weights
+
+In the early schema, the `dynamic_embedding.Variable` will directly be trained by optimizers and we had to extand all optimizers one-by-one.
+
+Now we propose a new scheme which no longer require to extand all optimizers one by one: 
+
+- The trainable warp class `dynamic_embdding.TrainableWrap` inherted from `ResourceVariable` will be introduced, and the main function is maintaining the relationship between `params` and `ids` created by `embedding_lookup` or `embedding_lookup_sparse`.
+
+```python
+from tensorflow.python.ops import resource_variable_ops
+
+class TrainableWrap(resource_variable_ops.ResourceVariable):
+  """
+  This class is a trainable warp of dynamic embedding,
+  and the key role is recording the map relationship of params and ids.
+  inheriting from the ResourceVariable make it trainable.
+  """
+  def __init__(self, params, ids, *args, **kwargs):
+    """Creates an empty `TrainableWrap` object.
+
+    Creates a group of tables placed on devices,
+    the type of its keys and values are specified by key_dtype
+    and value_dtype, respectively.
+
+    Args:
+      params: A dynamic_embedding.Variable instance.
+      ids: a tensor with any shape as same dtype of params.key_dtype.
+      ```other parameters is similar to tf.nn.embedding_lookup.```
+    Returns:
+      A `TrainableWrap` object which is a subclass of ResourceVariable.
+    """
+    self.params = params
+    self.ids = ids
+    self.prefetch_values = self.params.lookup(self.ids)
+
+    super(TrainableWrap, self).__init__(*args, **kwargs)
+
+  def update_op(self):
+    return self.params.upsert(self.ids, self._read_variable_op())
+```
+
+The scheme will **reuse** the `Optimizer._resource_apply_xxx` API to train `dynamic_embdding.TrainableWrap` naturally after some changing a little common code:
+
+- `ops/resource_variable_ops.py`: looking up the newest values to memory hold by `TrainableWrap->tensor()` before starting each iteration steps.
+- `training/optimizer.py`: creating the slots of `TrainableWarp` when primary `Variable` is an instance of `TrainableWarp`.
+- `training/optimizer.py`: writing back the trained weights of primary and slots to  hash tables of `dynamic_embedding.Variable` and after `Optimizer._resource_apply_xxx`.
+
+The pseudo-code is below: 
+
+```python
+# ops/resource_variable_ops.py:
+
+class BaseResourceVariable(variables.VariableV1):
+  """A python variable from an existing handle."""
+  def _read_variable_op(self):
+
+    # The `assign_variable_op` will be called to get before read opertion.
+    # `prefetch_values` is a lookup-like operation of `TrainableWrap`.
+
+    if hasattr(self, "prefetch_values"):
```

Comment:
Perhaps override `_read_variable_op`  instead of directly modifying base class?

Reply:
> Perhaps override `_read_variable_op` instead of directly modifying base class?

Ok, I fix it.